src/lib/png/png.c

```diff
@@ -0,0 +1,990 @@
+/**
+ * @file
+ * @brief lib for read and draw PNG
+ *
+ * @date May 10, 2019
+ * @author Filipp Chubukov
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <stdbool.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/ioctl.h>
+#include <lib/png.h>
+#include <fs/file_format.h>
+
+/* 256 literals, end code, length codes + 2 unused codes */
+#define NUMBER_OF_CODE_SYMBOLS 288
+
+/* the distance codes have their own symbols, 30 used, 2 unused */
+#define NUMBER_OF_DISTANCE_SYMBOLS 32
+
+/* lengths codes */
+#define NUMBER_OF_LENGTH_CODES 19
+#define FIRST_LENGTH_INDEX 257
+#define LAST_LENGTH_INDEX 285
+
+/**
+ * the base of distances(the bits of distance
+ * codes appear after length codes and use their own huffman tree)
+ */
+static const int DISTANCE_BASE[30]
+	= {1, 2, 3, 4, 5, 7, 9, 13, 17, 25, 33, 49, 65, 97, 129, 193, 257, 385, 513,
+		769, 1025, 1537, 2049, 3073, 4097, 6145, 8193, 12289, 16385, 24577};
+
+/* addit bits of distances (added to base)*/
+static const int DISTANCE_ADDITIONAL[30]
+	= {0, 0, 0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6, 7, 7, 8,
+		8, 9, 9, 10, 10, 11, 11, 12, 12, 13, 13};
+
+/* represented by codes 257-285 */
+static const int LENGTH_BASE[29]
+	= {3, 4, 5, 6, 7, 8, 9, 10, 11, 13, 15, 17, 19, 23, 27, 31, 35, 43, 51, 59,
+		67, 83, 99, 115, 131, 163, 195, 227, 258};
+
+/* addit bits with codes 257-285 (added to base length) */
+static const int LENGTH_ADDITIONAL[29]
+	= {0, 0, 0, 0, 0, 0, 0,  0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3,
+		4, 4, 4, 4, 5, 5, 5, 5, 0};
+/**
+ * the order in which "code length alphabet code lengths" are stored, out of this
+ * the huffman tree of the dynamic huffman tree lengths is generated
+ */
+static const int CLCL_ORDER[NUMBER_OF_LENGTH_CODES]
+	= {16, 17, 18, 0, 8, 7, 9, 6, 10, 5, 11, 4, 12, 3, 13, 2, 14, 1, 15};
+
+#define READBIT(bitpointer, bitstream) ((bitstream[bitpointer >> 3] >> (bitpointer & 0x7)) & (unsigned char)1)
+
+#define IHDR_END 33 /* position of end of IHDR chunk */
+#define IHDR_SIZE 13 /* len of IHDR data */
+
+struct png_header {
+	uint32_t width;
+	uint32_t height;
+
+	uint8_t bit_depth;
+	uint8_t color_type; /* 1 - plt, 2 - RGB, 4 or 6 - RGBA */
+	uint8_t compression_method; /* == 0 - deflate */
+	uint8_t filter_method; /* == 0 */
+	uint8_t interlace_method; /* 0 - standart, 1 - Adam7 */
+};
+
+/* Two functions to convert string of bytes to int */
+static int mypow(int digit, int degree) {
+	int i, res = 1;
+	for (i = 0; i < degree; i++) {
+		res = res * digit;
+	}
+	return res;
+}
+
+static int convert(uint8_t *bytes, int len) {
+	int i, res = 0;
```

Comment:
As far as I understand, this functions changes endianess of `*bytes`, and `len == 4`

There's `swab32` function for this in `swab.h`

Reply:
remove convert and add htonl from <netinet/in.h>